### PR TITLE
Optimize lighting calculation

### PR DESCRIPTION
**Pull request by TurkeyMcMac**
1. Node methods such as `getLight` can now be inlined.
2. Content lighting information is now stored in a fast cache. This is an array of size 2^16 embedded directly in `NodeDefManager`. Each entry is one byte. Because of its size, no bounds checks are needed when indexing it.

## To do

This PR is Ready for Review.

- [X] More testing
- [X] C++ benchmark

## How to test

Run the unit tests and try some mods.

Regarding performance, I have measured a ~20% decrease in the lighting benchmark's time.

### CODE REVIEW

File: src/mapnode.h [hunk 1]
```diff
@@ -71,6 +72,25 @@ typedef u16 content_t;
...
+*/
+struct ContentLightingFlags {
+	u8 light_source : 4;
+	bool has_light : 1;
+	bool light_propagates : 1;
+	bool sunlight_propagates : 1;
+
+	bool operator==(const ContentLightingFlags &other) const
+	{
+		return has_light == other.has_light && light_propagates == other.light_propagates &&
```
Comment (SmallJoker) [inline src/mapnode.h:86]: `memcmp` would be shorter

File: src/mapnode.h [hunk 2]
```diff
@@ -187,53 +207,97 @@ struct MapNode
...
+		// If node doesn't contain light data, ignore this
+		if (!f.has_light)
+			return;
+		if(bank == LIGHTBANK_DAY)
+		{
+			param1 &= 0xf0;
+			param1 |= a_light & 0x0f;
+		}
+		else if(bank == LIGHTBANK_NIGHT)
+		{
```
Comment (SmallJoker) [inline src/mapnode.h:221]: Style: Misplaced braces.

File: src/mapnode.h [hunk 3]
```diff
@@ -187,53 +207,97 @@ struct MapNode
...
+
+	u8 getLight(LightBank bank, ContentLightingFlags f) const noexcept
+	{
+		// Select the brightest of [light source, propagated light]
+		u8 light;
+		if(f.has_light)
+			light = bank == LIGHTBANK_DAY ? param1 & 0x0f : (param1 >> 4) & 0x0f;
+		else
+			light = 0;
+		return MYMAX(f.light_source, light);
```
Comment (SmallJoker) [inline src/mapnode.h:258]: Shorter, simpler.
```suggestion
	return MYMAX(f.light_source, getLightRaw(bank, f));
```
Also make `getLightRaw` inline.

File: src/mapnode.h [hunk 4]
```diff
@@ -187,53 +207,97 @@ struct MapNode
...
-	 * improvements in some situations.  Call this function if (and only if)
-	 * you have already retrieved the ContentFeatures by calling
-	 * NodeDefManager::get() for the node you're working with and the
-	 * pre-conditions listed are true.
-	 *
-	 * @pre f != NULL
-	 * @pre f->param_type == CPT_LIGHT
-	 */
-	u8 getLightNoChecks(LightBank bank, const ContentFeatures *f) const noexcept;
+	bool getLightBanks(u8 &lightday, u8 &lightnight, ContentLightingFlags f) const
```
Comment (SmallJoker) [inline src/mapnode.h:273]: Missing function description.

File: src/mapnode.h [hunk 5]
```diff
@@ -187,53 +207,97 @@ struct MapNode
...
+	{
+		// Select the brightest of [light source, propagated light]
+		if(f.has_light)
+		{
+			lightday = param1 & 0x0f;
+			lightnight = (param1>>4)&0x0f;
```
Comment (SmallJoker) [inline src/mapnode.h:279]: Wouldn't a bitfield struct be easier to deal with than manual shift operations every time?

File: src/client/mapblock_mesh.cpp [hunk 6]
```diff
@@ -126,17 +126,19 @@ u16 getInteriorLight(MapNode n, s32 increment, const NodeDefManager *ndef)
 static u8 getFaceLight(enum LightBank bank, MapNode n, MapNode n2,
 	v3s16 face_dir, const NodeDefManager *ndef)
 {
+	ContentLightingFlags f = ndef->getLightingFlags(n);
```
Comment (SmallJoker) [inline src/client/mapblock_mesh.cpp:129]: f1. Consistent naming.

File: src/dummygamedef.h [hunk 7]
```diff
@@ -0,0 +1,81 @@
+/*
```
Comment (SmallJoker) [inline src/dummygamedef.h:1]: These dummy files are used by the benchmark, so it might make sense to put them there?

File: src/nodedef.h [hunk 8]
```diff
@@ -819,6 +831,11 @@ class NodeDefManager {
 	 * Even constant NodeDefManager instances can register listeners.
 	 */
 	mutable std::vector<NodeResolver *> m_pending_resolve_callbacks;
+
+	/*!
+	 * Fast cache of content lighting flags.
+	 */
+	ContentLightingFlags m_content_lighting_flag_cache[CONTENT_MAX + 1L];
```
Comment (SmallJoker) [inline src/nodedef.h:838]: It is not necessary to waste RAM here. You could move these flags to `ContentFeatures` and implement an inline `NodeDefManager::getNoCheck(content_t id)` function to keep the lookup speed.

**Review by SmallJoker — COMMENTED** [review]
That's quite some changes for a 20% uplift...

**Comment by TurkeyMcMac** [comment]
The only change in this rebase is a simpler implementation of `isLightDayNightEq`.

**Comment by SmallJoker** [comment]
After going through the PR a few times, I believe that the performance lift-up is mostly caused by the following:
 * less lookups: reduction of getter calls (previously `ndef->get()`, now `ndef->getLightingFlags()`)
 * less indirection. `ContentFeatures` always needs dereferencing in the getter/setter functions. Having a struct that fits into a register (up to 8 bytes) avoids memory lookups
 * inline placement of crucial functions

I am concerned about the duplication of light data - the values should be present in either `ContentFeatures` or `NodeDefManager`, not in both to keep the code clean.

**Comment by TurkeyMcMac** [comment]
It's easier to have the lighting information in both the `ContentFeatures` and the cache. When a node is being registered, the lighting information is put into `ContentFeatures`. `NodeDefManager` then reads this information and puts it into its internal cache. Moving lighting information out of `ContentFeatures` would require redesigning this code, as well the code for serialization etc. As it is, `ContentFeatures::getLightingFlags` exists to keep the lighting cache in-sync. `ContentFeatures` structs within the `NodeDefManager` are only externally available immutably AFAICT, so there is not a risk of the cache becoming stale through external mutation.

**Comment by SmallJoker** [comment]
Yes, external mutation is not a problem here. I was thinking about code maintenance in the future. Luckily these values are yet not meant to change during runtime.
If separating is not easily possible, I think there should be at least a note in `ContentFeatures` for the relevant members so that the cache is not forgotten when there's a change in the light behavior/calculation.

**Comment by TurkeyMcMac** [comment]
I think the flags should be kept in a separate cache, since the overhead of of `vector::size()`, pointer dereferencing, and `string::empty()` could be significant with such a common operation. Perhaps I can run a test.

**Comment by TurkeyMcMac** [comment]
I have run the benchmarks on a x86_64 Mac from 2014 or so. I was testing this patch specifically:

```patch
diff --git a/src/nodedef.cpp b/src/nodedef.cpp
index e5f056807..30e1e3de6 100644
--- a/src/nodedef.cpp
+++ b/src/nodedef.cpp
@@ -1095,11 +1095,10 @@ void NodeDefManager::clear()
 		f.name = "unknown";
 		for (int t = 0; t < 6; t++)
 			f.tiledef[t].name = "unknown_node.png";
+		f.lighting_flags = f.getLightingFlags();
 		// Insert directly into containers
 		content_t c = CONTENT_UNKNOWN;
 		m_content_features[c] = f;
-		for (u32 ci = 0; ci <= CONTENT_MAX; ci++)
-			m_content_lighting_flag_cache[ci] = f.getLightingFlags();
 		addNameIdMapping(c, f.name);
 	}
 
@@ -1117,10 +1116,10 @@ void NodeDefManager::clear()
 		f.buildable_to        = true;
 		f.floodable           = true;
 		f.is_ground_content   = true;
+		f.lighting_flags      = f.getLightingFlags();
 		// Insert directly into containers
 		content_t c = CONTENT_AIR;
 		m_content_features[c] = f;
-		m_content_lighting_flag_cache[c] = f.getLightingFlags();
 		addNameIdMapping(c, f.name);
 	}
 
@@ -1137,10 +1136,10 @@ void NodeDefManager::clear()
 		f.diggable            = false;
 		f.buildable_to        = true; // A way to remove accidental CONTENT_IGNOREs
 		f.is_ground_content   = true;
+		f.lighting_flags      = f.getLightingFlags();
 		// Insert directly into containers
 		content_t c = CONTENT_IGNORE;
 		m_content_features[c] = f;
-		m_content_lighting_flag_cache[c] = f.getLightingFlags();
 		addNameIdMapping(c, f.name);
 	}
 }
@@ -1393,7 +1392,7 @@ content_t NodeDefManager::set(const std::string &name, const ContentFeatures &de
 		eraseIdFromGroups(id);
 
 	m_content_features[id] = def;
-	m_content_lighting_flag_cache[id] = def.getLightingFlags();
+	m_content_features[id].lighting_flags = def.getLightingFlags();
 	verbosestream << "NodeDefManager: registering content id \"" << id
 		<< "\": name=\"" << def.name << "\""<<std::endl;
 
@@ -1605,8 +1604,8 @@ void NodeDefManager::deSerialize(std::istream &is, u16 protocol_version)
 		// All is ok, add node definition with the requested ID
 		if (i >= m_content_features.size())
 			m_content_features.resize((u32)(i) + 1);
+		f.lighting_flags = f.getLightingFlags();
 		m_content_features[i] = f;
-		m_content_lighting_flag_cache[i] = f.getLightingFlags();
 		addNameIdMapping(i, f.name);
 		TRACESTREAM(<< "NodeDef: deserialized " << f.name << std::endl);
 
diff --git a/src/nodedef.h b/src/nodedef.h
index 044fdd8b2..c1af32f5c 100644
--- a/src/nodedef.h
+++ b/src/nodedef.h
@@ -306,6 +306,7 @@ struct ContentFeatures
 	/*
 		Cached stuff
 	 */
+	ContentLightingFlags lighting_flags;
 #ifndef SERVER
 	// 0     1     2     3     4     5
 	// up    down  right left  back  front
@@ -584,8 +585,7 @@ class NodeDefManager {
 	}
 
 	inline ContentLightingFlags getLightingFlags(content_t c) const {
-		// No bound check is necessary, since the array's length is CONTENT_MAX + 1.
-		return m_content_lighting_flag_cache[c];
+		return get(c).lighting_flags;
 	}
 
 	inline ContentLightingFlags getLightingFlags(const MapNode &n) const {
@@ -838,11 +838,6 @@ class NodeDefManager {
 	 * Even constant NodeDefManager instances can register listeners.
 	 */
 	mutable std::vector<NodeResolver *> m_pending_resolve_callbacks;
-
-	/*!
-	 * Fast cache of content lighting flags.
-	 */
-	ContentLightingFlags m_content_lighting_flag_cache[CONTENT_MAX + 1L];
 };
 
 NodeDefManager *createNodeDefManager();

```

Without patch:
* `voxalgo::update_lighting_nodes`: ~530 microseconds
* `voxalgo::blit_back_with_light`: ~1525 microseconds

With patch:
* `voxalgo::update_lighting_nodes`: ~530 microseconds
* `voxalgo::blit_back_with_light`: ~1660 microseconds

The time with and without the patch was about the same for `voxalgo::update_lighting_nodes`, but the patch made `voxalgo::blit_back_with_light` substantially slower.

**Comment by SmallJoker** [comment]
Thank you for the benchmark. I also did a rough in-game test with WorldEdit, adding a `TimerTaker` to `l_write_to_map`. The difference is surprisingly high.

```
//fixedpos set1 -60 45 150
//fixedpos set2 -100 40 90
//set default:brick
//set air
^ repeat a few times
```
PR: 16915 us (median)
Patch: 18934 us (median)  -> +12% execution time

With GCC 9.4.0 and release build, the functions are placed inline entirely, leaving less headroom for optimizations. We are in low millisecond range of functions that are not used too often. The users are unlikely to notice the difference, but faster code is nonetheless nice to have.

**Comment by TurkeyMcMac** [comment]
Alas, so it goes with caches.

**Review by SmallJoker — APPROVED** [review]
Works

**Pull request merged**